arangod/RestServer/DatabaseFeature.cpp

```diff
@@ -72,6 +71,17 @@ using namespace arangodb::basics;
 using namespace arangodb::options;
 
 namespace {
+
+template<typename T>
+void waitUnique(std::shared_ptr<T> const& ptr) {
+  // It's safe because we guarantee we have not weak_ptr
+  while (ptr.use_count() != 1) {
+    // 250ms from old DataProtector code authored by Max Neunhoeffer
+    std::this_thread::sleep_for(std::chrono::microseconds(250));
+  }
+  std::atomic_thread_fence(std::memory_order_acquire);
```

Comment:
What is this atomic-fence for? And why do we need waitUnique at all if we now use a shared_ptr for the database list?

Reply:
As I mentioned exists other approaches but I think this most close to old code

Reply:
Why do we need a HB relation between releasing the shared_ptr in read mode? A thread that is accessing the database list in read mode cannot change anything, so I don't see why we need to have a HB relation. And all changes to the list are protected by a mutex anyway.

Reply:
Because someone have access to databases pointers, then release shared pointer then we need unique access to this pointer

Reply:
Let's see example

t1:
read shared_ptr

t2:
lock mutex
update shared_ptr
move to vocbase ptr from shared_ptr list to dropped
// no wait or wait without acquire
unlock mutex

t3 -- gc thread:
lock mutex
access to dropped databese.


We want to t3 should be after t1 release it shared_ptr

Reply:
So I have chain:
t1 shared_ptr release synchronize with t2 wait
t2 mutex.unlock synchronize with t3 mutex.lock

Reply:
> In the old code we had to wait until it was ensured that no other thread had a reference to the memory, so it was safe to delete it.

It's pretty same, we have shared_ptr to **map of raw pointers**.
So wait in old and new code protect this raw pointers

Reply:
Discussed offline - the shared_ptr only protects the databases list, but we actually need to protect reclamation of the raw vocbase pointers from that list, which is why we need to wait until all other shared_ptrs are dropped.